modules/tensor_mechanics/include/materials/ADGeneralizedRadialReturnStressUpdate.h

```diff
@@ -0,0 +1,162 @@
+                                        const ADReal & residual,
+                                        const ADReal & scalar_effective_inelastic_strain) override;
+
+  virtual ADReal
+  minimumPermissibleValue(const ADDenseVector & /*effective_trial_stress*/) const override
+  {
+    return 0.0;
+  }
+
+  virtual ADReal
+  maximumPermissibleValue(const ADDenseVector & effective_trial_stress) const override;
+
+  /**
+   * Compute the limiting value of the time step for this material
+   * @return Limiting time step
+   */
+  virtual Real computeTimeStepLimit() override;
+
+  /**
+   * Does the model require the elasticity tensor to be isotropic?
+   */
+  bool requiresIsotropicTensor() override { return true; }
+
+protected:
+  virtual void initQpStatefulProperties() override;
+
+  /**
+   * Propagate the properties pertaining to this intermediate class.  This
+   * is intended to be called from propagateQpStatefulProperties() in
+   * classes that inherit from this one.
+   * This is intentionally named uniquely because almost all models that derive
+   * from this class have their own stateful properties, and this forces them
+   * to define their own implementations of propagateQpStatefulProperties().
+   */
+  void propagateQpStatefulPropertiesRadialReturn();
+
+  /**
+   * Perform any necessary initialization before return mapping iterations
+   * @param effective_trial_stress Effective trial stress
+   * @param elasticityTensor     Elasticity tensor
+   */
+  virtual void computeStressInitialize(const ADDenseVector & /*effective_trial_stress*/,
+                                       const ADRankFourTensor & /*elasticity_tensor*/)
+  {
+  }
+
+  /**
+   * Perform any necessary steps to finalize state after return mapping iterations
+   * @param inelasticStrainIncrement Inelastic strain increment
+   * @param delta_gamma Plastic multiplier
+   * @param stress Cauchy stress
+   * @param stress_dev Deviatoric part of Cauchy stress
+   */
+  virtual void computeStressFinalize(const ADRankTwoTensor & inelasticStrainIncrement,
+                                     const ADReal & delta_gamma,
+                                     ADRankTwoTensor & stress,
+                                     const ADDenseVector & /*stress_dev*/) = 0;
+  /**
+   * Perform any necessary steps to finalize strain increment after return mapping iterations
+   * @param inelasticStrainIncrement Inelastic strain increment
+   * @param stress Cauchy stress
+   * @param stress_dev Deviatoric part of Cauchy stress
+   * @param delta_gamma Plastic multiplier
+   */
+  virtual void computeStrainFinalize(ADRankTwoTensor & /*inelasticStrainIncrement*/,
+                                     const ADRankTwoTensor & /*stress*/,
+                                     const ADDenseVector & /*stress_dev*/,
+                                     const ADReal & /*delta_gamma*/)
+  {
+    mooseError("computeStrainFinalize needs to be implemented by a child class.");
```

Comment:
again, this should be pure virtual to make it a compile error and not a runtime error.